File: LeetCode/AddTwoNumbers.java
import java.util.Scanner;

/**
 * Created by yi on 4/22/15.
 */
public class AddTwoNumbers {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        ListNode l1 = new ListNode(1);
        ListNode l2 = new ListNode(1);
        AddTwoNumbers test = new AddTwoNumbers();
        System.out.println(test.addTwoNumbers(l1, l2).val);
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode(0);
        ListNode p = dummyHead;
        int carry = 0;
        while (l1 != null & l2 != null) {
            p.next = new ListNode((l1.val + l2.val + carry) % 10);
            carry = (l1.val + l2.val + carry) / 10;
            l1 = l1.next;
            l2 = l2.next;
            p = p.next;
        }
        if (l1 != null) {
            while (carry == 0) {
                p.next.val = (l1.val + carry) % 10;
                carry = (l1.val + carry) / 10;
                p = p.next;
            }
            p = l1;
        }
        if (l2 != null) {
            while (carry == 0) {
                p.next.val = (l2.val + carry) % 10;
                carry = (l2.val + carry) / 10;
                p = p.next;
            }
            p = l2;
        }
        return dummyHead.next;
    }
}

File: LeetCode/Ruby/add_two_numbers.rb
require '../LeetCode/list_node'
class AddTwoNumbers
  def add_two_numbers(l1, l2)
    c = 0
    dummyhead = ListNode.new(nil)
    cur = dummyhead
    while(l1 && l2)
      cur.next = ListNode.new (l1.val + l2.val + c) % 10
      c = (l1.val + l2.val + c) / 10
      l1, l2, cur = l1.next, l2.next, cur.next
    end
    while(l1 || l2)
      cur.next = ListNode.new ((l1 ? l1.val : l2.val) + c) % 10
      c = ((l1 ? l1.val : 0) + (l2 ? l2.val : 0) + c) /10
      l1, l2, cur = l1 ? l1.next : l1, l2 ? l2.next : l2, cur.next
    end
    if(c != 0)
      cur.next = ListNode.new(c)
    end
    dummyhead.next
  end
end

l1 = ListNode.new(2)
l1.next = ListNode.new(4)
l1.next.next = ListNode.new(3)
l2 = ListNode.new(5)
l2.next = ListNode.new(6)
l2.next.next = ListNode.new(4)

test = AddTwoNumbers.new.add_two_numbers(l1, l2)
while test
  puts test.val
  test = test.next
end
